src/mudlet.cpp

```diff
@@ -4122,6 +4122,553 @@ void mudlet::playSound(const QString& s, int soundVolume)
+            } else {
+                fileUrl = QUrl::fromUserInput(QStringLiteral("%1%2").arg(pHost->getMSPSoundLocation(), soundFileName));
+            }
+        }
+    }
+
+    return fileUrl;
+}
+
+bool mudlet::downloadMSPFile(QString& soundFileName, const QString& soundType, QString& absolutePathFileName)
+{
+    bool wasFileDownloaded = false;
+    QPointer<Host> pHost = getActiveHost();
+
+    if (pHost) {
+        QDir dir;
+        QString cacheDir = QStandardPaths::writableLocation(QStandardPaths::CacheLocation);
+        if (!dir.mkpath(cacheDir)) {
+            qWarning() << "mudlet::playMSPSound() WARNING - Couldn't create cache directory for sound file(s): " << cacheDir;
+            return wasFileDownloaded;
+        }
+
+        auto manager = new QNetworkAccessManager(this);
+        auto diskCache = new QNetworkDiskCache(this);
+        diskCache->setCacheDirectory(cacheDir);
+        manager->setCache(diskCache);
+
+        QUrl fileUrl = getMSPFileUrl(soundFileName, soundType);
+
+        if (!mudlet::self()->isMSPValidUrl(fileUrl)) {
+            return wasFileDownloaded;
+        } else {
+            QNetworkRequest request = QNetworkRequest(fileUrl);
+            request.setRawHeader(QByteArray("User-Agent"), QByteArray(QStringLiteral("Mozilla/5.0 (Mudlet/%1%2)").arg(APP_VERSION, APP_BUILD).toUtf8().constData()));
+            request.setAttribute(QNetworkRequest::FollowRedirectsAttribute, true);
+            request.setAttribute(QNetworkRequest::CacheLoadControlAttribute, QNetworkRequest::PreferCache);
+#ifndef QT_NO_SSL
+            if (fileUrl.scheme() == QStringLiteral("https")) {
+                QSslConfiguration config(QSslConfiguration::defaultConfiguration());
+                request.setSslConfiguration(config);
+            }
+#endif
+            pHost->updateProxySettings(manager);
+            QNetworkReply* getReply = manager->get(request);
+
+            connect(getReply, static_cast<void (QNetworkReply::*)(QNetworkReply::NetworkError)>(&QNetworkReply::error), this, [=](QNetworkReply::NetworkError) {
+                qWarning() << "mudlet::playMSPSound() WARNING - couldn't download sound from " << fileUrl.url();
+                getReply->deleteLater();
+            });
+
+            connect(getReply, &QNetworkReply::finished, this,
+                std::bind([=](QNetworkReply* reply) {
+                    if (reply->error() != QNetworkReply::NoError) {
+                        return wasFileDownloaded;
+                    }
+
+                    QFile file(absolutePathFileName);
+
+                    file.open(QFile::WriteOnly);
+                    file.write(reply->readAll());
+                    file.flush();
+                    file.close();
+
+                    reply->deleteLater();
+                }, getReply));
+
+            wasFileDownloaded = true;
+        }
+    }
+
+    return wasFileDownloaded;
+}
+
+QMediaPlayer* mudlet::getMSPSoundMediaPlayer()
```

Comment:
Given that you are creating/using two separate `QMediaPlayer` instances for sound effects and background music it might be useful to differentiate between them by classifying them with `QMediaPlayer::setAudioRole(QAudio::Role audioRole)` and possibly `QMediaPlayer::setCustomAudioRole(const QString &audioRole)` if `QAudio::CustomRole` is the one used...

This may help the end user to adjust the relative volumes of each as on some OS/Desktop Environments each separate `QMediaPlayer` gets listed in the system's Volume/Mixer controls and for long running one if they all claim to be just for the `Mudlet` application it might be difficult to tell them apart...

Reply:
I'll investigate supporting this.  I am not sure yet, but perhaps this could help differentiate between sound playing on different profiles as well.

Would we consider adding a button to the Main Toolbar to toggle Sound and Music?

Reply:
Let's consider that when we feel the need from the community - better to build things up that way.

Reply:
Enhanced!  As MSP audio is registered with a `QMediaPlayer`, Mudlet will find that:

- `QAudio::CustomRole` will return for `audioRole()` for the `QMediaPlayer` instance
- A `QString` with a custom pattern will return for `customAudioRole()` for the `QMediaPlayer` instance

The `QString` custom pattern returned by `customAudioRole()` is formatted like:

- For Sound - `QAudio::GameRole=<absolute_path_of_the_profiles_sound_directory>`
- For Music - `QAudio::MusicRole=<absolute_path_of_the_profiles_sound_directory>`

This pattern provides for the uniqueness necessary to stop sounds or music for the given profile per the MSP specification, while not interfering with other sound operations outside of the profile.

The pattern could be leveraged in future functionality to stop MSP sounds or music across all profiles, or other interesting things like tuning the volume of the sound or music beyond what was initially set.

Reply:
As mentioned previously, I took another approach, due to spending a couple days troubleshooting `QAudio` related code, and it seems to have worked out.